src/React/Basic.purs

```diff
@@ -27,115 +35,184 @@ instance semigroupJSX :: Semigroup JSX where
 instance monoidJSX :: Monoid JSX where
   mempty = empty
 
--- | A React component which can be used from JavaScript.
-foreign import data Component :: Type -> Type
+type ComponentSpec props state action =
+  { "$$type" :: ReactComponent props
+  , initialState :: state
+  , shouldUpdate :: LimitedSelf props state -> props -> state -> Boolean
+  , didMount :: Self props state action -> Effect Unit
+  , didUpdate :: Self props state action -> Effect Unit
+  , willUnmount :: LimitedSelf props state -> Effect Unit
+  , update :: Update props state action
+  , render :: Self props state action -> JSX
+  }
```

Comment:
It would be handy to have a couple of functions that allow changing the component type.

In our case we want to use the component both from JS and Purescript, we have this helper:

```haskell
contramapComponentProps
  :: forall state propsA propsB
   . (propsB -> propsA)
  -> ComponentSpec propsA state
  -> ComponentSpec propsB state
contramapComponentProps mapProps spec = spec
  { receiveProps = \self -> spec.receiveProps $ self
      { props = mapProps self.props }
  , render = \self -> spec.render $ self
      { props = mapProps self.props }
  }
```

which allows us to define our components with a pattern like:

```haskell
type Props =
  { foo :: Maybe Bar }

type JSProps =
  { foo :: Nullable Bar } 

fromJSProps :: JSProps -> Props

componentSpec :: ComponentSpec Props State

component :: Component Props
component = React.component componentSpec

jsComponent :: Component JSProps
jsComponent = React.component $ React.contramapComponentProps fromJSProps componentSpec
```

We haven't yet had a need for changing the `state` type (since state is internal), but I guess it might come useful (as well as changing the action type).

Reply:
That makes sense. In this PR, `Component` is an alias for `ComponentSpec` the `react-basic` specification of a component, not an actual React component. Because of that, this PR adds the `ReactComponent props` type and `toReactComponent :: ComponentSpec props _ _ -> ReactComponent`. Maybe a version of `toReactComponent` could provide this prop mapping you're doing.

Reply:
We could just change the type of `toReactComponent`, since `identity` is easy enough to pass in:
```purs
toReactComponent
  :: forall props state action jsProps
   . ({ | jsProps } -> props)
  -> ComponentSpec props state state action
  -> ReactComponent { | jsProps }
```

Reply:
This exists now

Reply:
@spicydonuts This is a lovely solution, thanks.